protocol/restlicodec/protobuf_writer.go

```diff
@@ -0,0 +1,319 @@
+	} else {
+		p.WriteOrdinal(pbDoubleOrdinal)
+		p.WriteVarDouble(v)
+	}
+}
+
+func (p *protobufWriter) WriteString(v string) {
+	if p.options.symbolTable != nil {
+		id, found := p.options.symbolTable.GetSymbolId(v)
+		if found {
+			p.WriteOrdinal(pbStringReferenceOrdinal)
+			p.WriteVarint(int64(id))
+			return
+		}
+	}
+	p.WriteOrdinal(pbStringLiteralOrdinal)
+	b := []byte(v)
+	p.WriteVarint(int64(len(b)))
+	p.buf.Write(b)
+}
+
+// WriteBytes implements PrimitiveWriter
+func (p *protobufWriter) WriteBytes(v []byte) {
+	p.WriteOrdinal(pbRawBytesOrdinal)
+	p.WriteVarint(int64(len(v)))
+	p.buf.Write(v)
+}
+
+// WriteOrdinal writes the given ordinal byte to the buffer
+func (p *protobufWriter) WriteOrdinal(v pbOrdinal) {
+	p.buf.WriteByte(byte(v))
+}
+
+func (p *protobufWriter) RawByte(v byte) {
+	p.buf.WriteByte(v)
+}
+func (p *protobufWriter) Raw(v []byte, _ error) {
+	p.buf.Write(v)
+}
+func (p *protobufWriter) RawString(v string) {
+	p.buf.WriteString(v)
+}
+func (p *protobufWriter) WriteRawBytes(data []byte) {
+	p.Raw(data, nil)
+}
+func (p *protobufWriter) BuildBytes(...[]byte) ([]byte, error) {
+	return p.buf.Bytes(), nil
+}
+func (p *protobufWriter) ReadCloser() io.ReadCloser {
+	return ioutil.NopCloser(bytes.NewReader(p.buf.Bytes()))
+}
+func (p *protobufWriter) Size() int {
+	return p.buf.Len()
+}
+func (p *protobufWriter) Finalize() string {
+	data, _ := p.BuildBytes()
+	return string(data)
+}
+
+//  https://developers.google.com/protocol-buffers/docs/encoding#varints
+func (p *protobufWriter) WriteVarint(v int64) {
+	if p.options.expVarintImpl {
+		// use the embedded varint implementation
+		p.writeVarintImpl(uint64(v))
+	} else {
+		// use the Go built-in varint implementation
+		var buf []byte = make([]byte, 4)
+		s := binary.PutVarint(buf, 4)
+		p.WriteRawBytes(buf[:s])
+	}
+}
+func (p *protobufWriter) WriteUvarint(v uint64) {
+	if p.options.expVarintImpl {
+		// use the embedded varint implementation
+		p.writeVarintImpl(v)
+	} else {
+		// use the Go built-in varint implementation
+		var buf []byte = make([]byte, 16)
+		s := binary.PutUvarint(buf, 16)
+		p.WriteRawBytes(buf[:s])
+	}
+}
+func (p *protobufWriter) WriteVarDouble(v float64) {
+	uintBits := math.Float64bits(v)
+	p.WriteUvarint(uintBits)
+}
+func (p *protobufWriter) writeFixed32(value uint32) {
+	p.buf.WriteByte(byte(value & 0xFF))
+	p.buf.WriteByte(byte((value >> 8) & 0xFF))
+	p.buf.WriteByte(byte((value >> 16) & 0xFF))
+	p.buf.WriteByte(byte((value >> 24) & 0xFF))
+}
+func (p *protobufWriter) writeFixed64(value uint64) {
+	p.buf.WriteByte(byte(value & 0xFF))
+	p.buf.WriteByte(byte((value >> 8) & 0xFF))
+	p.buf.WriteByte(byte((value >> 16) & 0xFF))
+	p.buf.WriteByte(byte((value >> 24) & 0xFF))
+	p.buf.WriteByte(byte((value >> 32) & 0xFF))
+	p.buf.WriteByte(byte((value >> 40) & 0xFF))
+	p.buf.WriteByte(byte((value >> 48) & 0xFF))
+	p.buf.WriteByte(byte((value >> 56) & 0xFF))
+}
+func (p *protobufWriter) writeVarintImpl(value uint64) {
+
+}
```

Comment:
A custom Varint write implementation goes here, if `encoding/binary.PutVarint` isn't correct.

Behavior currently toggled via `ProtobufOptions.expVarintImpl`, but eventually we should only use whichever implementation is compatible with restli java. Currently the `encoding/binary` is always used.

Still unsure if a custom Varint implementation is necessary, so I didn't implement a similar behavior toggle on the Reader side, but it is trivial to add.